lib/zclient.c

```diff
@@ -3036,6 +3036,114 @@ static void zclient_mlag_handle_msg(ZAPI_CALLBACK_ARGS)
 		(*zclient->mlag_handle_msg)(zclient->ibuf, length);
 }
 
+/*
+ * Send an OPAQUE message, contents opaque to zebra. The message header
+ * is a message subtype.
+ */
+int zclient_send_opaque(struct zclient *zclient, uint32_t type,
+			const uint8_t *data, size_t datasize)
+{
+	int ret;
+	struct stream *s;
+
+	/* Check buffer size */
+	if (STREAM_SIZE(zclient->obuf) <
+	    (ZEBRA_HEADER_SIZE + sizeof(type) + datasize))
+		return -1;
+
+	s = zclient->obuf;
+	stream_reset(s);
+
+	zclient_create_header(s, ZEBRA_OPAQUE_MESSAGE, VRF_DEFAULT);
+
+	/* Send sub-type */
+	stream_putl(s, type);
+
+	/* Send opaque data */
+	stream_write(s, data, datasize);
+
```

Comment:
I think we need to encode Opaque message with TLV (Type Length Value) in particular if we would send a batch of Opaque messages (particular useful when re-sync TED between IS-IS and BGP for example). In addition, there is a risk to miss information or going out-of-bound when reading Opaque message on the receiver side. So, I would suggest to add
```
/* Send Opaque data size */
stream_putl(datasize);
```
Just before sending the opaque data

Reply:
certainly the receiver should be careful about parsing the payload, but the overall size of the payload is readily available - the header holds the overall size, as does the incoming 'stream' object at the receiver. I think it's best to keep the opaque message body ... opaque?

Reply:
Agree that Opaque is Opaque, but, looking to IETF usual practices, Opaque is encoded within a TLV structure. Here, Length is missing even if you could retrieve the information in the Header. The idea behind Opaque, is that the data will be structured by TLV encoding a.k.a. Russian doll model. Having a top TLV in the message body will ease decoding without looking into the ZAPI header structure. In addition, if we would batch many Opaque information, we could insert several Opaque TLV in the same ZAPI message. Here, the size in the Zapi header will provide information about the number of Opaque TLV is embed in the Opaque message. Of course the type of Opaque Message must be the same.

Reply:
Gosh, look: the whole point of 'opaque' is that it's opaque. the only things that are known to the zapi infrastructure are the zapi message type, and the first four octets. inside that, you can ... invent tlvs, or you can plop protobuf, or memcpy structs, or anything else your application needs. whether you choose to include one application-level object, or a series of objects, or a nested complex of objects is entirely up to your application. zapi has no business trying to decode or understand that information - that's the responsibility of the applications at the endpoints.

as a general point, I agree with you that tlvs are generally a good thing - I've written several RFCs taking that approach. but that's not what this PR is doing - it's explicitly avoiding forcing a decision about the encoding of the payloads of the OPAQUE messages. that's an application matter. the payloads that emerge for the TED/LU use-case may well be completely different from the ones that Lou B uses in the use-case that he's thought about, for example.

Reply:
OK. Let's have this TLV encoding for the sub type and we'll change that if needed.